samples/ChangeFeedProcessor/DocumentDB.ChangeFeedProcessor.Test/IntegrationTests/IntegrationTest.cs

```diff
@@ -0,0 +1,208 @@
+            {
+                lock (this.ClassData.testContextSyncRoot)
+                {
+                    TestContext.Properties[leaseCollectionInfoPropertyName] = value;
+                }
+            }
+        }
+
+        protected TestClassData ClassData
+        {
+            get { return testClasses[this.GetType().Name]; }
+        }
+
+        [AssemblyInitialize]
+        public static void AssemblyInitialize(TestContext testContext)
+        {
+            System.Net.ServicePointManager.DefaultConnectionLimit = 1000;  // Default is 2.
+            ThreadPool.SetMinThreads(1000, 1000);   // 32
+            ThreadPool.SetMaxThreads(5000, 5000);   // 32
+        }
+
+        [AssemblyCleanup]
+        public static async Task AssemblyCleanupAsync()
+        {
+            var endpoint = new Uri(ConfigurationManager.AppSettings["endpoint"]);
+            var masterKey = ConfigurationManager.AppSettings["masterKey"];
+            var databaseName = ConfigurationManager.AppSettings["databaseId"];
+
+            using (var client = new DocumentClient(endpoint, masterKey))
+            {
+                await client.DeleteDatabaseAsync(UriFactory.CreateDatabaseUri(databaseName));
+            }
+        }
+
+        [TestInitialize]
+        public async Task TestInitializeAsync()
+        {
+            if (this.ClassData.monitoredCollectionInfo == null)
+            {
+                await this.ClassData.classInitializeSyncRoot.WaitAsync(); // Use semaphore as cannot await inside a lock.
+
+                try
+                {
+                    if (this.ClassData.monitoredCollectionInfo == null)
+                    {
+                        this.ClassData.leaseCollectionInfoTemplate = await TestClassInitializeAsync(this, $"data_{this.GetType().Name}");
+                    }
+                }
+                finally
+                {
+                    this.ClassData.classInitializeSyncRoot.Release();
+                }
+            }
+
+            this.LeaseCollectionInfo = new DocumentCollectionInfo(this.ClassData.leaseCollectionInfoTemplate);
+            this.LeaseCollectionInfo.CollectionName = $"leases_{this.GetType().Name}_{Guid.NewGuid().ToString()}";
+
+            var leaseCollection = new DocumentCollection { Id = this.LeaseCollectionInfo.CollectionName };
+            using (var client = new DocumentClient(this.LeaseCollectionInfo.Uri, this.LeaseCollectionInfo.MasterKey, this.LeaseCollectionInfo.ConnectionPolicy))
+            {
+                await Helper.CreateDocumentCollectionAsync(client, this.LeaseCollectionInfo.DatabaseName, leaseCollection, leaseOfferThroughput);
+            }
+        }
+
+        [TestCleanup]
+        public async Task TestCleanupAsync()
+        {
+            Debug.Assert(this.LeaseCollectionInfo != null);
+            using (var client = new DocumentClient(this.LeaseCollectionInfo.Uri, this.LeaseCollectionInfo.MasterKey, this.LeaseCollectionInfo.ConnectionPolicy))
```

Comment:
Just a thought but what if the Base Class contains a ConcurrentDictionary of Clients instead of creating and `using` them in several places, with a method that lets the Test classes that inherit from the base Get clients from these Dictionary? If no client exists on request (we can use the Uri as key), we create one and add to Dictionary.

Reply:
Too many DocumentClient instances may cause perf issues. We'd like to minimize the number of clients kept for long time..